### PR TITLE
Run tests with sanitizers and the debug allocator, use PyMem_Raw instead of malloc-family

### PR DESCRIPTION
**Pull request by michaelfm1211**
- [X] I hereby agree to dual licence this and any previous contributions under both
the _Biopython License Agreement_ **AND** the _BSD 3-Clause License_.

- [X] I have read the ``CONTRIBUTING.rst`` file, have run ``pre-commit``
locally, and understand that continuous integration checks will be used to
confirm the Biopython unit tests and style checks pass with these changes.

- [X] I have added my name to the alphabetical contributors listings in the files
``NEWS.rst`` and ``CONTRIB.rst`` as part of this pull request, am listed
already, or do not wish to be listed. (*This acknowledgement is optional.*)

Run the GitHub CI Linux tests with the [address sanitizer](https://clang.llvm.org/docs/AddressSanitizer.html) and the [undefined behavior sanitizer](https://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html). Additionally, the [Python debug allocator](https://docs.python.org/3/c-api/memory.html#debug-hooks-on-the-python-memory-allocators) is enabled during the test.

Calls to the `malloc`-family of C standard library functions have been replaced by calls to the [`PyMem_Raw` family of functions](https://docs.python.org/3/c-api/memory.html#raw-memory-interface) so the debug allocator can track them during testing.

These changes will help catch bugs in the C code that might be introduced in the future. It should not change anything noticeable to the user.

### CODE REVIEW

**Comment by michaelfm1211** [comment]
Huh, odd. It looks like both the debug allocator and ASAN are working, but the debug allocator is catching something in Bio/Cluster/clustermodule.c.. This may or may not be Biopython's fault, I'll like into it later. It could also be ASAN and the debug allocator tripping over each other's checks. The debug allocator's error message says to expect segfaults because of overwritten memory used by the allocator, the ASAN error is probably superfluous.

**Comment by michaelfm1211** [comment]
It looks like the errors are coming from allocating memory with C's `malloc` and `calloc` but then later freeing it with `PyMem_Free`. This is specifically mentioned as disallowed by the [Memory Management docs](https://docs.python.org/3/c-api/memory.html#allocator-domains):
> When freeing memory previously allocated by the allocating functions belonging to a given domain,the matching specific deallocating functions must be used. For example, [PyMem_Free()](https://docs.python.org/3/c-api/memory.html#c.PyMem_Free) must be used to free memory allocated using [PyMem_Malloc()](https://docs.python.org/3/c-api/memory.html#c.PyMem_Malloc).

I'll add a commit soon to fix these discrepancies. It's probably best to replace all calls to `malloc` and `calloc` with `PyMem_RawMalloc` and `PyMem_RawCalloc` at the least so that the debug allocator can track them during testing.

File: Bio/Cluster/cluster.c
Comment (mdehoon) [inline Bio/Cluster/cluster.c:1]: Bio/Cluster/cluster.c and Bio/Cluster/clustermodule.c are also used outside of Biopython. For this reason, we cannot simply replace malloc etc. with PyMem_malloc etc. But you could have a `#define` that defines `MALLOC` to be either `malloc` or `PyMem_malloc`, and then use `MALLOC` in the code.

Comment (michaelfm1211) [reply Bio/Cluster/cluster.c:1]: Ok good to know, thanks! Bio/Cluster/cluster.c doesn't interact with Python on its own in any way, so we could leave it to use `malloc` then in Bio/Cluster/clustermodule.c just create a new `PyMem_Malloc` allocation, copy whatever `treecluster` returns into the new Python allocation, then free the old `malloc` allocation. I added a new commit to do this.

I think it's a bit cleaner than changes ~200 lines of code in Bio/Cluster/cluster.c and the copying only gets done once per call to `treecluster`. The code in Bio/Cluster/cluster.c won't be checked with the debug allocator, but that's fine because ASAN will still check it. It doesn't interact with Python so there's no much use of having the debug allocator check it anyway.

Thoughts on this? If you think this is a bad approach I can implement the macro system you suggested instead.

**Comment by peterjc** [comment]
Interesting work - good to see :)

File: Bio/Cluster/cluster.c
Comment (mdehoon) [reply Bio/Cluster/cluster.c:1]: Thank you. I think your approach works, but I would prefer using macros in Bio/Cluster/cluster.c. Using the PyMem_* instead of malloc/calloc/free allows Python to keep track of how much memory is being used, and call the garbage collector at the appropriate time. The changes in Bio/Cluster/cluster.c are just simple replacements, and is not error-prone.

Comment (michaelfm1211) [reply Bio/Cluster/cluster.c:1]: Ok, sounds good. I've pushed a new commit that uses the PyMem_* family instead in Bio/Cluster/cluster.c if `CLUSTER_USE_PYTHON_MEMORY` is defined at compile time, and changed the setup.py so Biopython compiles the extension with `-DCLUSTER_USE_PYTHON_MEMORY`. Please share any thoughts on this or the new code.

File: Bio/Cluster/cluster.c
Comment (mdehoon) [inline Bio/Cluster/cluster.c:44]: Can you change this to
```C
#ifdef CLUSTER_USE_PYTHON_MEMORY
#include "Python.h"
#define MALLOC PyMem_RawMalloc
#define CALLOC PyMem_RawCalloc
#define REALLOC PyMem_RawRealloc
#define FREE PyMem_RawFree
#else
#define MALLOC malloc
#define CALLOC calloc
#define REALLOC realloc
#define FREE free
#endif
```
as I am not sure if it is safe to just redefine malloc etc., and also the code should remain readable to people who do not use Python.
Then, you can just replace all instances of malloc/calloc/realloc/free by their equivalent in uppercase.

Comment (mdehoon) [reply Bio/Cluster/cluster.c:49]: Code looks good now, thanks!

**Review by mdehoon — APPROVED** [review]

**Comment by mdehoon** [comment]
No I think it is fine; I already gave my approval above. @JoaoRodrigues  is the remaining pending reviewer, but I am not sure if he is actively following this PR.

**Comment by peterjc** [comment]
@MarkusPiotrowski can you give a second opinion?

This has gone beyond my C comfort zone, but seems reasonable for the CI builds.

I would say go ahead and merge this after allowing a week for comment.

**Comment by MarkusPiotrowski** [comment]
> @MarkusPiotrowski can you give a second opinion?
> 
> This has gone beyond my C comfort zone, but seems reasonable for the CI builds.

Sorry, I can't. Also beyond my C comfort zone.

**Comment by peterjc** [comment]
OK - let's merge it then. Thanks Michael!

**Pull request merged**
